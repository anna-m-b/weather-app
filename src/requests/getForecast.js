/* eslint-disable no-console */
import axios from 'axios';

const getForecast = (
  setSelectedDate,
  setLocation,
  setForecasts,
  setErrorCode,
  city
) => {
  const path = city
    ? `https://mcr-codes-weather-app.herokuapp.com/forecast?city=${city}`
    : `https://mcr-codes-weather-app.herokuapp.com/forecast`;
  axios
    .get(path)
    .then((res) => {
      setSelectedDate(res.data.forecasts[0].date);
      setLocation(res.data.location);
      setForecasts(res.data.forecasts);
    })
    .catch((error) => {
      if (error.response) {
        setErrorCode(error.response.status);
      } else if (error.request) {
        setErrorCode(500);
        console.log(error.request);
      } else {
        console.log('Error', error.message);
      }
      console.log(error.config);
    });
};

export default getForecast;
