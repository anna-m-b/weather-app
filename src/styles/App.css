.App {
  padding: 1rem;
  background: #56CCF2;  /* fallback for old browsers */
  background: -webkit-linear-gradient(to top, #2F80ED, #56CCF2);  /* Chrome 10-25, Safari 5.1-6 */
  background: linear-gradient(to top, #2F80ED, #56CCF2); /* W3C, IE 10+/ Edge, Firefox 16+, Chrome 26+, Opera 12+, Safari 7+ */
  /* display: flex;
  flex-direction: column;
  align-content: center; */
}

/* .App > * {
  margin: 1rem;
} */

.Loading {
  margin: 1rem;
  font-size: 1.5rem;
  color: black;
}