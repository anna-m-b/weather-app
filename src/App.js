/* eslint-disable no-unused-vars */
/* eslint-disable no-console */
import './styles/App.css';
import React, { useState, useEffect } from 'react';
import WeeklyForecast from './containers/WeeklyForecast';
import Location from './components/Location';
import DetailedForecast from './components/DetailedForecast';
import getForecast from './requests/getForecast';
import SearchForm from './components/SearchForm';
import ErrorMessage from './components/ErrorMessage';
import Title from './components/Title';

function App() {
  const [location, setLocation] = useState({});
  const [forecasts, setForecasts] = useState([]);
  const [selectedDate, setSelectedDate] = useState();
  const [searchInput, setSearchInput] = useState('');
  const [city, setCity] = useState('');
  const [errorCode, setErrorCode] = useState();

  const selectedForecast = forecasts.find(
    (forecast) => forecast.date === selectedDate
  );

  useEffect(() => {
    getForecast(setSelectedDate, setLocation, setForecasts, setErrorCode, city);
  }, [city]);

  useEffect(() => {
    if (errorCode) {
      setLocation({});
      setSelectedDate();
    }
  }, [errorCode]);

  const handleMoreDetailsClick = (date) => {
    setSelectedDate(date);
  };

  const onInputChange = (e) => {
    setSearchInput(e.target.value);
  };

  const onSubmit = () => {
    setErrorCode();
    setCity(searchInput);
    setSearchInput('');
  };

  if (!forecasts.length) {
    return <h1 className="Loading">Loading...</h1>;
  }

  return (
    <div className="App">
      <Title />
      <SearchForm
        onInputChange={onInputChange}
        onSubmit={onSubmit}
        searchInput={searchInput}
      />
      <Location location={location} />

      {errorCode && <ErrorMessage errorCode={errorCode} city={city} />}

      {selectedForecast && <DetailedForecast forecastItem={selectedForecast} />}
      <WeeklyForecast
        forecasts={forecasts}
        handleMoreDetailsClick={handleMoreDetailsClick}
        errorCode={errorCode}
      />
    </div>
  );
}

export default App;
