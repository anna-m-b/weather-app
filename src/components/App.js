import '../styles/App.css';
import React from 'react';
import PropTypes from 'prop-types';
import WeeklyForecast from './containers/WeeklyForecast';
import Location from './components/Location';
import DetailedForecast from './components/DetailedForecast';
import '../data/forecast.json';

function App({ forecastData }) {
  const { location } = forecastData;

  return (
    <div className="App">
      <h1>Weather Forecast</h1>
      <Location location={location} />
      <WeeklyForecast />
      <DetailedForecast />
    </div>
  );
}

App.propTypes = {
  forecastData: PropTypes.shape({
    location: PropTypes.shape({
      city: PropTypes.string,
      country: PropTypes.string,
    }),
    forecasts: PropTypes.arrayOf(PropTypes.any),
  }).isRequired,
};

export default App;
