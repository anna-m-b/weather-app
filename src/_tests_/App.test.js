import React from 'react';
import { render, waitFor } from '@testing-library/react';
import App from '../App';

import getForecast from '../requests/getForecast';

jest.mock('../requests/getForecast');

describe('App', () => {
  beforeEach(() => {
    getForecast.mockResolvedValue({
      data: {
        location: {
          city: 'Manchester',
          country: 'UK',
        },
        forecasts: [
          {
            date: 1525046400000,
            temperature: {
              max: 11,
              min: 4,
            },
            wind: {
              speed: 13,
              direction: 's',
              icon: 'wi wi-wind towards-180-deg',
            },
            humidity: 30,
            description: 'Clear',
            icon: 800,
          },
          {
            date: 1525132800000,
            temperature: {
              max: 13,
              min: 8,
            },
            wind: {
              speed: 60,
              direction: 'ne',
              icon: 'wi wi-wind towards-90-deg',
            },
            humidity: 80,
            description: 'Stormy',
            icon: 211,
          },
        ],
      },
    });
  });

  it('shows loading text while waiting for response on first load', () => {
    getForecast.mockReturnValue(new Promise(() => {}));
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly after getting data', async () => {
    const { asFragment, getByText } = render(<App />);
    await waitFor(() => expect(getByText(/manchester/i)).toBeInTheDocument());
    expect(asFragment()).toMatchSnapshot();
  });
});
