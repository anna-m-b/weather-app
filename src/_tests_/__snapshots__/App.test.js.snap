// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App shows loading text while waiting for response on first load 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h1
      class="Location"
    >
      No city selected
    </h1>
    <div
      class="SearchForm"
    >
      <label
        for="search"
      >
        <input
          id="search"
          placeholder="Enter a city"
          type="text"
          value=""
        />
      </label>
      <button
        data-testid="search-button"
        type="submit"
      >
        Search
      </button>
    </div>
    <div
      class="WeeklyForecast"
    />
  </div>
</DocumentFragment>
`;
